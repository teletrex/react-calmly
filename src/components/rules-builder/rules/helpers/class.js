/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import classNames from 'classnames';
import map from 'lodash/map';
import isArray from 'lodash/isArray';
import  settings  from '../../../../settings';

const { prefix } = settings;

export const getPrefixedClasses = (classes = []) =>
  classNames(map(isArray(classes) ? classes : [classes], className => `${prefix}${className}`));
