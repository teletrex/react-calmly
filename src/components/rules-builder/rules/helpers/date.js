/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




const getDate = timestamp => (timestamp ? new Date(timestamp) : new Date());

export const getDayTimestamp = datestring => {
  const day = getDate(datestring);
  return day.setHours(0, 0, 0, 0);
};

export const getTodayTimestamp = () => getDayTimestamp();

export const getReadOnlyDate = timestamp => {
  const date = getDate(timestamp);
  return `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()}`;
};

// used for the datetime attribute in <time> tags
export const getMachineDateTime = timestamp => {
  const date = getDate(timestamp);
  return date.toISOString();
};
