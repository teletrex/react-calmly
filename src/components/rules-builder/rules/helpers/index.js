/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export * from './class';
export * from './date';
export * from './rule';
export * from './useFormValidation';
