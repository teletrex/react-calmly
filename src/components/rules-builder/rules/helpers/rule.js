/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import capitalize from 'lodash/capitalize';
import uniqueId from 'lodash/uniqueId';

import { composeOperators } from '../../utils/composeOperators';

export const getRuleOutput = (operator, value) => ({ operator, value });

export const getInitialOperator = (operatorValue, operators) =>
  operators.find(item => item.value === operatorValue);

export const getRuleConfigData = ruleConfig => {
  const { placeholder = {}, inputComponent } = ruleConfig;

  return {
    placeholder,
    ...(inputComponent && { inputComponent }),
  };
};

export const getRuleContainerConfigData = (ruleConfig, ruleData, operatorsList) => {
  const { disableCriteriaReselect, operators: ruleOperators = [] } = ruleConfig;
  const { value } = ruleData;
  const operators = ruleOperators.length
    ? composeOperators(ruleOperators, operatorsList)
    : operatorsList;
  let operator = getInitialOperator(ruleData.operator, operators);

  if (disableCriteriaReselect && !operator) {
    [operator] = operators;
  }

  return {
    operator,
    value,
    operators,
  };
};

export const getRuleId = ruleType => `${capitalize(ruleType)}Rule_${uniqueId()}`;
