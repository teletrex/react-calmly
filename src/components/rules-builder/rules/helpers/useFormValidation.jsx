/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { useFormik } from 'formik';

export function useFormValidation({ initialValues, onSubmit, validate, validationSchema }) {
  return useFormik({
    initialValues,
    onSubmit,
    validate: validate || undefined,
    validateOnBlur: true,
    validateOnChange: false,
    validationSchema,
  });
}
