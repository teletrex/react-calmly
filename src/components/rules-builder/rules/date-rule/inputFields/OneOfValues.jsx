/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { Fragment, useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import map from 'lodash/map';
import uniq from 'lodash/uniq';
import without from 'lodash/without';

import { DatePickerInput, Tag, DatePicker } from '@carbon/react';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { DATE_PATTERN } from '../../constants';
import { getPrefixedClasses } from '../../helpers';

const parseDate = timestamp => new Date(timestamp).toDateString();

const OneOfValues = ({ id, onValueChange, value }) => {
  const { isReadOnly } = useContext(RulesBuilderContext);
  const [timestamps, setTimestamps] = useState(value);

  useEffect(() => {
    setTimestamps(value);
  }, [value]);

  const addValue = dateValue => {
    const date = new Date(dateValue[0]);
    const newTimestamps = uniq([...timestamps, date.getTime()]);

    setTimestamps(newTimestamps);
    onValueChange(newTimestamps);
  };

  const removeValue = timestamp => {
    const newTimestamps = without(timestamps, timestamp);

    setTimestamps(newTimestamps);
    onValueChange(newTimestamps);
  };

  if (isReadOnly) {
    return map(timestamps, (timestamp, i) => (
      <Fragment key={timestamp}>
        <span>{i ? ',' : ':'}</span>
        <span className={getPrefixedClasses(['--text-space-left', '--rule-value-readonly'])}>
          {parseDate(timestamp)}
        </span>
      </Fragment>
    ));
  }

  return (
    <div
      className={getPrefixedClasses('--input-field-container')}
      data-testid="date-rule-one-of-values"
    >
      {map(timestamps, timestamp => (
        <Tag
          key={timestamp}
          filter
          onClick={() => {
            removeValue(timestamp);
          }}
          title={timestamp.toString()}
        >
          {parseDate(timestamp)}
        </Tag>
      ))}
      <DatePicker datePickerType="single" id={`${id}--date-rule-picker-1`} onChange={addValue}>
        <DatePickerInput id={`${id}--date-rule-value-1`} labelText="" pattern={DATE_PATTERN} />
      </DatePicker>
    </div>
  );
};

OneOfValues.propTypes = {
  id: PropTypes.string.isRequired,
  onValueChange: PropTypes.func.isRequired,
  value: PropTypes.arrayOf(PropTypes.number),
};

OneOfValues.defaultProps = {
  value: [],
};

export default OneOfValues;
