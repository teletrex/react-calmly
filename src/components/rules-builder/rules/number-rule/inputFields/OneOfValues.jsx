/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import map from 'lodash/map';
import uniq from 'lodash/uniq';
import  {useTranslation}  from '../../../../../translation';
import AddTags from '../../../../add-tags';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { PROPTYPES } from '../../constants';
import { getPrefixedClasses } from '../../helpers';

const OneOfValues = ({ id, onValueChange, operator, placeholder, value }) => {
  const { t } = useTranslation();
  const { isReadOnly } = useContext(RulesBuilderContext);

  const [valueTags, setValueTags] = useState(map(value, String));
  const [inputValue, setInputValue] = useState('');

  useEffect(() => {
    setInputValue('');
  }, [operator]);

  useEffect(() => {
    setValueTags(map(value, String));
  }, [value]);

  const handleAddTag = tags => {
    const newTags = uniq(map(tags, tag => tag.trim()));

    setValueTags(newTags);
    onValueChange(map(newTags, Number));
    setInputValue('');
  };

  const handleDeleteTag = tags => {
    setValueTags(tags);
    onValueChange(map(tags, Number));
  };

  const handleInputChange = newInputValue => {
    setInputValue(newInputValue);
  };

  if (isReadOnly) {
    return (
      <span className={getPrefixedClasses(['--rule-value-readonly', '--text-space-left'])}>
        {valueTags.join(', ')}
      </span>
    );
  }

  return (
    <AddTags
      allowTagSplitting={false}
      id={id}
      inline
      inlineInputAlwaysOn
      inputValue={inputValue}
      min={0}
      name={id}
      onAddTag={handleAddTag}
      onDelete={handleDeleteTag}
      onInputChange={handleInputChange}
      placeholder={placeholder || t('Enter value')}
      shapeForm="oval"
      tagExistsError={t('This value already exists')}
      tags={valueTags}
      type="number"
      validateInputValue
    />
  );
};

OneOfValues.propTypes = {
  id: PropTypes.string.isRequired,
  onValueChange: PropTypes.func.isRequired,
  operator: PROPTYPES.OPERATOR,
  placeholder: PropTypes.string,
  value: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.number, PropTypes.string])),
};

OneOfValues.defaultProps = {
  operator: null,
  placeholder: '',
  value: [],
};

export default OneOfValues;
