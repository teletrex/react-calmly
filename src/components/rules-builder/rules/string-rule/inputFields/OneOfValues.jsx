/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import map from 'lodash/map';
import uniq from 'lodash/uniq';

import AddTags from '../../../../add-tags';
import  {useTranslation}  from '../../../../../translation';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { PROPTYPES } from '../../constants';
import { getPrefixedClasses } from '../../helpers';

const OneOfValues = ({ id, onValueChange, operator, placeholder, value }) => {
  const { t } = useTranslation();
  const { isReadOnly } = useContext(RulesBuilderContext);

  const [valueTags, setValueTags] = useState(value);
  const [inputValue, setInputValue] = useState('');

  useEffect(() => {
    setInputValue('');
  }, [operator]);

  useEffect(() => {
    setValueTags(map(value, String));
  }, [value]);

  const handleAddTag = tags => {
    const newTags = uniq(map(tags, tag => tag.trim()));

    setValueTags(newTags);
    onValueChange(newTags);
    setInputValue('');
  };

  const handleDeleteTag = tags => {
    setValueTags(tags);
    onValueChange(tags);
  };

  const handleInputChange = newInputValue => {
    setInputValue(newInputValue);
  };

  if (isReadOnly) {
    return (
      <span className={getPrefixedClasses(['--rule-value-readonly', '--text-space-left'])}>
        {valueTags.join(', ')}
      </span>
    );
  }

  return (
    <AddTags
      allowTagSplitting={false}
      id={id}
      inline
      inlineInputAlwaysOn
      inputValue={inputValue}
      name={id}
      onAddTag={handleAddTag}
      onDelete={handleDeleteTag}
      onInputChange={handleInputChange}
      placeholder={placeholder || t('Enter value')}
      tagExistsError={t('This value already exists')}
      tags={valueTags}
      validateInputValue
    />
  );
};

OneOfValues.propTypes = {
  id: PropTypes.string.isRequired,
  onValueChange: PropTypes.func.isRequired,
  operator: PROPTYPES.OPERATOR,
  placeholder: PropTypes.string,
  value: PropTypes.arrayOf(PropTypes.string),
};

OneOfValues.defaultProps = {
  operator: null,
  placeholder: '',
  value: [],
};

export default OneOfValues;
