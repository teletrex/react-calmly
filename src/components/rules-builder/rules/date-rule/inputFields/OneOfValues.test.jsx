/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { getDayTimestamp } from '../../helpers';
import { datePickerErrorWrapper, renderWithContext } from '../../../utils/testHelpers';

import OneOfValues from './OneOfValues';

const initialDate = '11/11/2020';
const initialValue = [getDayTimestamp(initialDate)];

const props = {
  id: 'TestDateRuleElement',
  onValueChange: jest.fn(),
};

describe('OneOfValues', () => {
  it('should render correctly', () => {
    const { container } = render(<OneOfValues {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a provided value', () => {
    datePickerErrorWrapper(() => {
      const { container } = render(<OneOfValues {...props} value={initialValue} />);
      expect(container).toMatchSnapshot();
    });
  });

  it('should render correctly for read-only', () => {
    // add a second value to test read-only conditions for colon and comma:
    const initialValues = initialValue.concat([getDayTimestamp('12/12/2020')]);
    const { container } = renderWithContext(<OneOfValues {...props} value={initialValues} />, {
      isReadOnly: true,
    });

    expect(container).toMatchSnapshot();
  });

  it('should set the rule value when a date is entered to date picker', () => {
    datePickerErrorWrapper(() => {
      const { getByPlaceholderText } = render(<OneOfValues {...props} />);

      const datePickerInput = getByPlaceholderText('MM/DD/YYYY');

      fireEvent.change(datePickerInput, { target: { value: initialDate } });
      fireEvent.blur(datePickerInput);

      expect(props.onValueChange).toHaveBeenCalledWith(initialValue);
    });
  });

  it('should correctly remove an existing value', () => {
    datePickerErrorWrapper(() => {
      const { getByTitle } = render(<OneOfValues {...props} value={initialValue} />);
      const removeButton = getByTitle('Remove');

      fireEvent.click(removeButton);

      expect(removeButton).not.toBeInTheDocument();
    });
  });
});
