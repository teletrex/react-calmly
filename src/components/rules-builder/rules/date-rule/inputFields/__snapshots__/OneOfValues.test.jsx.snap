// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneOfValues should render correctly 1`] = `
<div>
  <div
    class="cds--input-field-container"
    data-testid="date-rule-one-of-values"
  >
    <div
      class="cds--form-item"
    >
      <div
        class="cds--date-picker cds--date-picker--single"
        id="TestDateRuleElement--date-rule-picker-1"
      >
        <span
          class="cds--date-picker-input__wrapper--relative"
        >
          <div
            class="cds--date-picker-container cds--date-picker--nolabel"
          >
            <div
              class="cds--date-picker-input__wrapper"
            >
              <input
                autocomplete="off"
                class="cds--date-picker__input flatpickr-input"
                id="TestDateRuleElement--date-rule-value-1"
                pattern="d{1,2}/d{1,2}/d{4}"
                placeholder="MM/DD/YYYY"
                type="text"
                value=""
              />
              <svg
                aria-label="Open calendar"
                class="cds--date-picker__icon"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13,2h-2V1h-1v1H6V1H5v1H3C2.4,2,2,2.4,2,3v10c0,0.6,0.4,1,1,1h10c0.6,0,1-0.4,1-1V3C14,2.4,13.6,2,13,2z M13,13H3V6h10V13z M13,5H3V3h2v1h1V3h4v1h1V3h2V5z"
                />
                <title>
                  Open calendar
                </title>
              </svg>
            </div>
          </div>
          <svg
            aria-hidden="true"
            class="cds--date-picker-input__error cds--date-picker-input__error--hidden"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 1C4.1 1 1 4.1 1 8s3.1 7 7 7 7-3.1 7-7-3.1-7-7-7zm2.7 10.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
            />
            <path
              d="M10.7 11.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
              data-icon-path="inner-path"
              fill="none"
              opacity="0"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`OneOfValues should render correctly for read-only 1`] = `
<div>
  <span>
    :
  </span>
  <span
    class="cds--text-space-left cds--rule-value-readonly"
  >
    Wed Nov 11 2020
  </span>
  <span>
    ,
  </span>
  <span
    class="cds--text-space-left cds--rule-value-readonly"
  >
    Sat Dec 12 2020
  </span>
</div>
`;

exports[`OneOfValues should render correctly with a provided value 1`] = `
<div>
  <div
    class="cds--input-field-container"
    data-testid="date-rule-one-of-values"
  >
    <div
      class="cds--tag cds--tag--gray cds--tag--filter cds--tag--oval"
      title=""
    >
      <div
        class="cds--tag__label"
      >
        <span
          class="cds--tag__text-ellipsis"
        >
          Wed Nov 11 2020
        </span>
      </div>
      <button
        aria-label="Remove"
        title="Remove"
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          style="will-change: transform;"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 4.7l-.7-.7L8 7.3 4.7 4l-.7.7L7.3 8 4 11.3l.7.7L8 8.7l3.3 3.3.7-.7L8.7 8z"
          />
        </svg>
      </button>
    </div>
    <div
      class="cds--form-item"
    >
      <div
        class="cds--date-picker cds--date-picker--single"
        id="TestDateRuleElement--date-rule-picker-1"
      >
        <span
          class="cds--date-picker-input__wrapper--relative"
        >
          <div
            class="cds--date-picker-container cds--date-picker--nolabel"
          >
            <div
              class="cds--date-picker-input__wrapper"
            >
              <input
                autocomplete="off"
                class="cds--date-picker__input flatpickr-input"
                id="TestDateRuleElement--date-rule-value-1"
                pattern="d{1,2}/d{1,2}/d{4}"
                placeholder="MM/DD/YYYY"
                type="text"
                value=""
              />
              <svg
                aria-label="Open calendar"
                class="cds--date-picker__icon"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13,2h-2V1h-1v1H6V1H5v1H3C2.4,2,2,2.4,2,3v10c0,0.6,0.4,1,1,1h10c0.6,0,1-0.4,1-1V3C14,2.4,13.6,2,13,2z M13,13H3V6h10V13z M13,5H3V3h2v1h1V3h4v1h1V3h2V5z"
                />
                <title>
                  Open calendar
                </title>
              </svg>
            </div>
          </div>
          <svg
            aria-hidden="true"
            class="cds--date-picker-input__error cds--date-picker-input__error--hidden"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 1C4.1 1 1 4.1 1 8s3.1 7 7 7 7-3.1 7-7-3.1-7-7-7zm2.7 10.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
            />
            <path
              d="M10.7 11.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
              data-icon-path="inner-path"
              fill="none"
              opacity="0"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
